exec/java-exec/src/main/java/org/apache/drill/exec/compile/JaninoClassCompiler.java

```diff
@@ -58,5 +56,27 @@ public JaninoClassCompiler(ClassLoader parentClassLoader, boolean debug) {
   }
 
   @Override
+  public Map<String,byte[]> compile(final ClassNames className, final String sourceCode)
+      throws CompileException, IOException, ClassNotFoundException {
+
+    ClassFile[] classFiles = doCompile(sourceCode);
+    Map<String,byte[]> results = new HashMap<>();
+    for(int i = 0;  i < classFiles.length;  i++) {
+      ClassFile classFile = classFiles[i];
+      results.put(classFile.getThisClassName(), classFile.toByteArray());
```

Comment:
As we've discussed, this is the class loader, not the global class cache. The non-ASM path can't get the class name from the byte codes, so the compiler has to tell us, via this map, which byte code corresponds to which compiled class name. The "plain old Java" class loader then fetches classes given their compiled names using this map.